src/vs/editor/common/model/treeSitterTokens.ts

```diff
@@ -0,0 +1,226 @@
+		const parser = await this._getParser();
+		const newLanguage = this._languageId();
+		const [language, queries] = await Promise.all([this._getLanguage(newLanguage), this._getQueries(newLanguage)]);
+		parser.setLanguage(language);
+		if (!this._tokens || this._lastLanguageId !== newLanguage) {
+			this._tokens?.dispose();
+			this._tokens = new TextModelTokens(this._textModel, parser, queries, language, this._colorThemeData);
+		}
+
+	}
+
+	private async _getParser(): Promise<Parser> {
+		if (!this._parser) {
+			await this._treeSitterService.initTreeSitter();
+			this._parser = new Parser();
+		}
+		return this._parser;
+	}
+
+	private async _getLanguage(newLanguage: string): Promise<Parser.Language> {
+		if (!this._language || this._lastLanguageId !== newLanguage) {
+			this._language = await this._fetchLanguage();
+		}
+		return this._language;
+	}
+
+	private async _fetchLanguage() {
+		const grammarName = TreeSitterTokenizationRegistry.get(this._languageId());
+		const wasmPath: AppResourcePath = `vs/base/common/treeSitterLanguages/${grammarName?.name}/${grammarName?.name}.wasm`;
+		const languageFile = await (this._fileService.readFile(FileAccess.asFileUri(wasmPath)));
+		return Parser.Language.load(languageFile.value.buffer);
+	}
+
+	private async _getQueries(newLanguage: string): Promise<string> {
+		if (!this._queries || this._lastLanguageId !== newLanguage) {
+			this._queries = await this._fetchQueries(newLanguage);
+		}
+		return this._queries;
+	}
+
+	private async _fetchQueries(newLanguage: string): Promise<string> {
+		const grammarName = TreeSitterTokenizationRegistry.get(newLanguage);
+		const scmPath: AppResourcePath = `vs/base/common/treeSitterLanguages/${grammarName?.name}/highlights.scm`;
+		const query = await this._fileService.readFile(FileAccess.asFileUri(scmPath));
+		return query.value.toString();
+	}
+
+
+	public getLineTokens(lineNumber: number): LineTokens {
+		if (this._tokens) {
+			return new LineTokens(this._tokens.lineTokens(lineNumber), this._textModel.getLineContent(lineNumber), this._languageIdCoded);
+		}
+		return LineTokens.createEmpty('', this._languageIdCoded);
+	}
+
+	public resetTokenization(fireTokenChangeEvent: boolean = true): void {
+		// TODO @alexr00 determine if needed
+	}
+
+}
+
+
+/**
+ * For handling the text model changes.
+ */
+class TextModelTokens extends Disposable {
+	private _tree: Parser.Tree | undefined;
+	private _query: Parser.Query | undefined;
+
+	constructor(private readonly _textModel: TextModel,
+		private readonly _parser: Parser,
+		private readonly _queries: string,
+		private readonly _language: Parser.Language,
+		private readonly _colorThemeData: ColorThemeData) {
+		super();
+		this._register(this._textModel.onDidChangeContent(() => this._onDidChangeContent()));
+	}
+
+	private _onDidChangeContent() {
+		// TODO @alexr00 this is an extremely naive implementation and should absolutely not be pushed to main, it's just a placeholder.
+		this._tree?.delete();
```

Comment:
@hediet this is just temporary. I will implement proper content change handling in an upcoming commit.